### ISSUE TITLE
ResNetC4Model

### ISSUE THREAD

**Issue by chenxiaodanhit** [tensorpack/tensorpack#1063]
Why is the input to the RPN network only feature map without anchors ?
Did I get it wrong?
`

  def rpn(self, image, features, inputs):#创建rpn网络,输入为原图、feature map和anchors的类别和bounding box的位置
        featuremap = features[0]
        rpn_label_logits, rpn_box_logits = rpn_head('rpn', featuremap, cfg.RPN.HEAD_DIM, cfg.RPN.NUM_ANCHOR) #创建rpn的分类和回归的卷积层 dim=1024,anchors=15,15种不同的anchors
       `

**Comment by ppwwyyxx** [comment]
RPN does use anchors and you can find the word "anchors" in the code.

**Comment by chenxiaodanhit** [comment]
Which line of code?
Let me explain a little bit more, I've spent couple of days read your code and I have two questions:

1.  Does RPN only use resnet's conv4 when calculating label and box in this function? I didn't find any anchors in this line 'rpn_label_logits, rpn_box_logits = rpn_head('rpn', featuremap, cfg.RPN.HEAD_DIM, cfg.RPN.NUM_ANCHOR) '. And I thought this line constructed the basic model of RPN head.

def rpn_head(featuremap, channel, num_anchors): #featuremap 1024 15
    """
    Returns:
        label_logits: fHxfWxNA
        box_logits: fHxfWxNAx4
    """
    with argscope(Conv2D, data_format='channels_first',
                  kernel_initializer=tf.random_normal_initializer(stddev=0.01)):
        hidden = Conv2D('conv0', featuremap, channel, 3, activation=tf.nn.relu) #1024个3*3的卷积核对resnet的输出conv4进行卷积

        label_logits = Conv2D('class', hidden, num_anchors, 1) #创建分类的卷积网络
        box_logits = Conv2D('box', hidden, 4 * num_anchors, 1) #创建回归的卷积网络
        # 1, NA(*4), im/16, im/16 (NCHW)

        label_logits = tf.transpose(label_logits, [0, 2, 3, 1])  # 1xfHxfWxNA
        label_logits = tf.squeeze(label_logits, 0)  # fHxfWxNA

        shp = tf.shape(box_logits)  # 1x(NAx4)xfHxfW
        box_logits = tf.transpose(box_logits, [0, 2, 3, 1])  # 1xfHxfWx(NAx4)
        box_logits = tf.reshape(box_logits, tf.stack([shp[2], shp[3], num_anchors, 4]))  # fHxfWxNAx4
    return label_logits, box_logits

2. I'm also wondering the function of the code below.
Is this code only make a little adjustment to the RPN bounding box?

pred_boxes_decoded = anchors.decode_logits(rpn_box_logits)

I'm a 菜鸟 and thanks for your time!

**Comment by ppwwyyxx** [comment]
> I didn't find any anchors in this line

Because anchors are not needed in this particular line and is needed later.

I recommend you to read the series of object detection papers instead. In general we do not answer machine learning questions in tensorpack issues. This is a place to learn to use the tensorpack library, not a place to learn machine learning models.
